### PR TITLE
Remove the web service

### PR DESCRIPTION
**Pull request by TristanCacqueray**
This change simplifies monocle operation by replacing the nginx web proxy
with a wai middleware, so that the webui files are directly served by the API.

### CODE REVIEW

**Comment by morucci** [comment]
That looks already great ! Good to see one service less in the deployment !

**Comment by TristanCacqueray** [comment]
I think this is now ready, the k8s deployment should work after the images are published.

File: Dockerfile-api [hunk 1]
```diff
@@ -24,7 +24,33 @@ ARG MONOCLE_COMMIT
...
+# install dependencies
+RUN npm install
+
+# copy source files
+COPY web/build.js /monocle-webapp/
+COPY web/public /monocle-webapp/public/
+COPY web/src /monocle-webapp/src/
+
+# install dependencies (sed is ugly hack to make warning errors)
+RUN sed -e 's|-bs-no-version-header|-bs-no-version-header", "-warn-error -a+5+6+27+101+109|' -i bsconfig.json && npm run build
```
Comment (morucci) [inline Dockerfile-api:50]: This command is failing with:
Error: /webapp/src: No such file or directory

File: Dockerfile-api [hunk 2]
```diff
@@ -24,7 +24,33 @@ ARG MONOCLE_COMMIT
...
+COPY web/bsconfig.json /webapp/
+COPY web/package-lock.json /webapp/
+
+# install dependencies
+RUN npm install
+
+# copy source files
+COPY web/build.js /monocle-webapp/
+COPY web/public /monocle-webapp/public/
+COPY web/src /monocle-webapp/src/
```
Comment (morucci) [inline Dockerfile-api:47]: Perhaps copy should be in /webapp/ (see the error below)

**Review by morucci — COMMENTED** [review]
I detected some issues, please see inline comments.

File: haskell/src/Monocle/Main.hs [hunk 3]
```diff
@@ -21,24 +24,70 @@ import Network.Wai.Middleware.Cors (cors, corsRequestHeaders, simpleCorsResource
...
+          Nothing -> index
+
+    -- Helper that checks if `fp` exists, otherwise it returns `otherFP`
+    existOr :: FilePath -> Maybe FilePath -> IO (Maybe FilePath)
+    existOr fp otherFP = do
+      exist <- System.Directory.doesPathExist fp
+      pure $ if exist then Just fp else otherFP
+
+    -- The middleware pass the request to the monocle app
+    staticMiddleware :: LByteString -> FilePath -> Wai.Application -> Wai.Application
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply haskell/src/Monocle/Main.hs:71]: Not sure why it is not the case, perhaps we need to do it manually with https://hackage.haskell.org/package/wai-logger-2.4.0/docs/Network-Wai-Logger.html#t:ServerPushLogger

File: Dockerfile-codegen [hunk 4]
```diff
@@ -1,39 +0,0 @@
-# Monocle.
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply Dockerfile-codegen:1]: I guess that was already the case, the `codegen-with-container` was not documented :)

**Comment by morucci** [comment]
Next try it fails on the docker image build:
```
Step 10/19 : COPY web/bsconfig.json /webapp/
 ---> 3d55a7e34642
Step 11/19 : COPY web/package-lock.json /webapp/
 ---> 9092599083b5
Step 12/19 : RUN npm install
 ---> Running in e713bbe3ce32
npm ERR! code ENOENT
npm ERR! syscall open
npm ERR! path /monocle-webapp/package.json
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, open '/monocle-webapp/package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2022-06-02T17_31_15_402Z-debug-0.log
The command '/bin/sh -c npm install' returned a non-zero code: 254
ERROR: Service 'api' failed to build : Build failed
```

**Review by morucci — APPROVED** [review]
Looks good now !

**Pull request merged**